### PR TITLE
Чаусов - Лабораторная работа №2

### PR DESCRIPTION
**Pull request by ChausovND**

### CODE REVIEW

File: modules/chausov_n_bit_array/src/bit_array.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+#include <string>
+
+BitArray::BitArray() {
+  arr = NULL;
+  size = 0;
+  bitSize = 0;
+}
+
+BitArray::BitArray(const unsigned int _bitSize) {
+  size = _bitSize / 8 + 1;
```
Comment (dkurt) [inline modules/chausov_n_bit_array/src/bit_array.cpp:14]: What if `_bitSize` is divided by 8? Do we have an extra array element?

File: modules/chausov_n_bit_array/src/bit_array.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  bitSize = _bitSize;
+  arr = new char[size];
```
Comment (dkurt) [inline modules/chausov_n_bit_array/src/bit_array.cpp:16]: Use `std::vector<char>` or create a destructor which will clean memory.

File: modules/chausov_n_bit_array/src/bit_array.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  for (int i = 0; i < static_cast<int>(size); i++) arr[i] = 0;
```
Comment (dkurt) [inline modules/chausov_n_bit_array/src/bit_array.cpp:17]: Why not have `i` of type `unsigned int`?
```cpp
for (unsigned int i = 0; i < size; i++) arr[i] = 0;
```

File: modules/chausov_n_bit_array/src/bit_array.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  arr[pos] = arr[pos] | (static_cast<int>(val) << (7 - bitPos % 8));
+}
+
+bool BitArray::GetBit(unsigned int bitPos) {
+  if (bitPos >= size * 8) throw std::string("Out of range");
+  int pos = bitPos / 8;
+  return arr[pos] & 1 << (7 - bitPos % 8);
+}
+
+bool BitArray::IsEmpty() { return size == 0 && bitSize == 0 && arr == NULL; }
```
Comment (dkurt) [inline modules/chausov_n_bit_array/src/bit_array.cpp:41]: I think that it's enough check only `size`

File: modules/chausov_n_bit_array/src/bit_array.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
+// Copyright 2022 Chausov Nikita
+
+#include "include/bit_array.h"
+
+#include <string>
+
+BitArray::BitArray() {
+  arr = NULL;
```
Comment (dkurt) [inline modules/chausov_n_bit_array/src/bit_array.cpp:8]: `nullptr`

File: modules/chausov_n_bit_array/src/bit_array.cpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  if (bitPos >= size * 8) throw std::string("Out of range");
+  int pos = bitPos / 8;
+  return arr[pos] & 1 << (7 - bitPos % 8);
+}
+
+bool BitArray::IsEmpty() { return size == 0 && bitSize == 0 && arr == NULL; }
+
+void BitArray::DeleteArray() {
+  if (arr == NULL) throw std::string("Array alrady NULL");
+  for (int i = 0; i < static_cast<int>(size); i++) arr[i] = 0;
```
Comment (dkurt) [inline modules/chausov_n_bit_array/src/bit_array.cpp:45]: not needed

File: modules/chausov_n_bit_array/src/bit_array.cpp (hunk 1 above)
Comment (ChausovND) [reply modules/chausov_n_bit_array/src/bit_array.cpp]: Added a check for whole division

File: modules/chausov_n_bit_array/src/bit_array.cpp (hunk 2 above)
Comment (ChausovND) [reply modules/chausov_n_bit_array/src/bit_array.cpp:20]: Added a destructor

File: modules/chausov_n_bit_array/src/bit_array.cpp (hunk 3 above)
Comment (ChausovND) [reply modules/chausov_n_bit_array/src/bit_array.cpp]: Fixed

File: modules/chausov_n_bit_array/src/bit_array.cpp (hunk 4 above)
Comment (ChausovND) [reply modules/chausov_n_bit_array/src/bit_array.cpp:51]: Done

File: modules/chausov_n_bit_array/src/bit_array.cpp (hunk 6 above)
Comment (ChausovND) [reply modules/chausov_n_bit_array/src/bit_array.cpp]: Done

File: modules/chausov_n_bit_array/src/bit_array.cpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  int pos = bitPos / 8;
+  return arr[pos] & 1 << (7 - bitPos % 8);
+}
+
+bool BitArray::IsEmpty() { return size == 0; }
+
+void BitArray::DeleteArray() {
+  if (arr == nullptr) throw std::string("Array alrady NULL");
+
+  this->~BitArray();
```
Comment (dkurt) [inline modules/chausov_n_bit_array/src/bit_array.cpp:57]: Destructor should not be called explicitly. Better to call `DeleteArray` inside destructor

Comment (ChausovND) [reply modules/chausov_n_bit_array/src/bit_array.cpp]: Done

**Pull request merged**